pykwasm/src/pykwasm/kdist/wasm-semantics/binary-parsing/elem.md

```diff
@@ -0,0 +1,197 @@
+
+  syntax BinaryElemMode ::= #binaryElemActive( table: Int, offset: BinaryInstrs )
+  syntax BinaryDefnElem ::= #binaryElem
+            ( type: RefValType
+            , elemSegment: ListRef
+            , mode: BinaryElemMode
+            )
+  syntax BinaryDefn ::= BinaryDefnElem
+
+endmodule
+
+module BINARY-PARSER-ELEM  [private]
+  imports BINARY-PARSER-CONSTANT-SYNTAX
+  imports BINARY-PARSER-ELEM-SYNTAX
+  imports BINARY-PARSER-EXPR-SYNTAX
+  imports BINARY-PARSER-EXPR-VEC-SYNTAX
+  imports BINARY-PARSER-INT-SYNTAX
+  imports BINARY-PARSER-REFTYPE-SYNTAX
+  imports BINARY-PARSER-TAGS
+
+  rule parseDefn(defnElem, BWI:BytesWithIndex) => parseDefnElem(BWI)
+
+  syntax DefnResult ::= parseDefnElem(BytesWithIndex)  [function, total]
+                      | #parseDefnElem(IntResult)  [function, total]
+  rule parseDefnElem(BWI:BytesWithIndex) => #parseDefnElem(parseLeb128UInt(BWI))
+  rule #parseDefnElem(intResult(ELTS_ACTIVE_ZERO_BY_REF, BWI:BytesWithIndex)) => parseElem0(BWI)
+  rule #parseDefnElem(intResult(ELTS_PASSIVE_BY_REF, BWI:BytesWithIndex)) => parseElem1(BWI)
+  rule #parseDefnElem(intResult(ELTS_ACTIVE_IDX_BY_REF, BWI:BytesWithIndex)) => parseElem2(BWI)
+  rule #parseDefnElem(intResult(ELTS_DECL_BY_REF, BWI:BytesWithIndex)) => parseElem3(BWI)
+  rule #parseDefnElem(intResult(ELTS_ACTIVE_ZERO_BY_VAL, BWI:BytesWithIndex)) => parseElem4(BWI)
+  rule #parseDefnElem(intResult(ELTS_PASSIVE_BY_VAL, BWI:BytesWithIndex)) => parseElem5(BWI)
+  rule #parseDefnElem(intResult(ELTS_ACTIVE_IDX_BY_VAL, BWI:BytesWithIndex)) => parseElem6(BWI)
+  rule #parseDefnElem(intResult(ELTS_DECL_BY_VAL, BWI:BytesWithIndex)) => parseElem7(BWI)
+  rule #parseDefnElem(intResult(I:Int, BWI:BytesWithIndex)) => parseError("#parseDefnElem", ListItem(I) ListItem(BWI))
+      [owise]
+  rule #parseDefnElem(E:ParseError) => E
+
+  syntax DefnResult ::= parseElem0(BytesWithIndex)  [function, total]
+                      | #parseElem0a(ExprResult)  [function, total]
+                      | #parseElem0b(BinaryInstrs, IntsResult)  [function, total]
+  rule parseElem0(BWI:BytesWithIndex) => #parseElem0a(parseExpr(BWI))
+  rule #parseElem0a(exprResult(Is:BinaryInstrs, BWI:BytesWithIndex))
+      => #parseElem0b(Is, parseUnsignedIntVec(BWI))
+  rule #parseElem0a(E:ParseError) => E
+  rule #parseElem0b(E:BinaryInstrs, intsResult(L:Ints, BWI:BytesWithIndex))
+      => defnResult
+          ( #binaryElem
+              (... type: funcref
+              , elemSegment: intsToRefFunc(L)
+              , mode: #binaryElemActive(... table: 0, offset: E)
+              )
+          , BWI
+          )
+  rule #parseElem0b(_, E:ParseError) => E
+
+
+  syntax DefnResult ::= parseElem1(BytesWithIndex)  [function, total]
+                      | #parseElem1a(BytesWithIndexOrError)  [function, total]
+                      | #parseElem1b(IntsResult)  [function, total]
+  rule parseElem1(BWI:BytesWithIndex) => #parseElem1a(parseConstant(BWI, b"\x00"))
```

Comment:
This is a small nit; I would prefer to our existing macro for this constant to help readability a bit.

Also, right now, the macro is called `ELT_KIND`, but I think it might be better to call it `ELT_KIND_FUNCREF` --- I didn't pay close enough attention to realize that element kind reduces to a type.

The same applies to the other uses of the zero byte.

Reply:
Done.